DESIGN.md

```diff
@@ -0,0 +1,95 @@
+-  Trivial deployment.
+-  Monitor performance characteristics.
+-  Verbose logging to central ElasticSearch instance.
+-  Alerts:
+   -  Process start/finish with result and auxiliary information in case of
+      failure.
+   -  Big subprocesses' start/finish.
+
+## Wanted features
+
+-  _Pull rather than push data from Landonline._ This way we can keep the entire
+   system together, rather than having a costly and awkward split between
+   on-premise and cloud architecture.
+-  _Continuous deployment._ Any time a change is merged to the main branch some
+   smoke tests are performed and the result is deployed to production.
+-  _React to change rather than run on a timer,_ where possible. Ideally the BDE
+   Processor should know about new files as soon as the last file has been
+   closed for writing, and should immediately start processing.
+-  _From-scratch deployment friendly._ That is, we should always have the option
+   to deploy from scratch, deleting _all_ the resources we've created and
+   getting back to an identical state (from the end users' perspective) in a
+   timely fashion (that is, before the next data dump).
+-  _Access for everybody at the org._ This should be simpler than the current
+   solution, and should be easy with read replicas.
+-  _Public access._ This would require a separate user with access to only the
+   appropriately licensed tables (some have a restrictive customer license),
+   using a separate read-only replica.
+-  _Minimal code changes for upstream schema changes._ For example, rather than
+   using complex schema migration code, we should be able to just change the ETL
+   code and redeploy.
+-  Read replica use _does not affect the performance of the pipeline to LDS._
+   Basically just means we'll need to use separate read replicas for them.
+-  Zero configuration. The service should rely on Azure SSO so we don't need to
+   maintain users, we should not need to do any IP whitelisting, and the current
+   version of the BDE Processor should only support the current version of
+   Landonline.
+
+## Non-requirements
+
+-  _Don't keep BDE files around._ They should ideally be kept only in memory or
+   at most in temporary storage which is cleaned up after each run. If we
+   absolutely need to keep them around that should be done as a separate system.
+   If it's convenient, the BDE Processor could read from the data in that system
+   rather than doing its own data copying. TBC who needs the files, and whether
+   _someone_ needs to keep doing this.
+-  _No table versioning._ It creates a _lot_ of extra complexity, and we don't
+   have any PostgreSQL experts at hand to maintain the extensions.
+   -  TBD: Review how versioned tables are currently being used, and work out a
+      reasonable solution.
+
+## Data flow
+
+```mermaid
+graph LR
+   subgraph Landonline
+   informix[(Informix)] -->|daily| bde-repository[bde-repository]
```

Comment:
The 'bde-repository' should probably in it's own subgraph box. It's not really part of 'Landonline'. They only keep 1 or 2 weeks worth of files. So maybe could have a separate thing that keeps a 'repository' of files?

Reply:
"bde-repository" is the host name that BDE Transport connects to in order to download files; isn't that part of Landonline?

Reply:
Not sure I would call it a 'repositor'y. More like it dumps a daily file and then deletes ones older than a week or so, so a 'file share'?

I see further down you suggest no keeping old BDE files in AWS, which is more like a 'repository'. Let's resolve that I guess.

Reply:
Cool, I've renamed the entry in the graph to "file share".